### PR TITLE
Recipe App

### PR DESCRIPTION
**Pull request by elarhadu**
## Hi Code Reviewer 👋
For this milestone, my partner [Houda](@houdamzari) and I created a Recipe App with RoR. The project requirements were to create;
- [x] Login page and registration page:
   - Should be built with Devise.
- [x] Food list (only for a group that has less than 3 members):
   - Should display a list of food added by the logged-in user as in the wireframe (display also quantity of a given food).
   - Should lead to a form that allows users to add new food.
- [x] Recipes list:
   - Should display a list of recipes created by the logged-in user as in the wireframe.
   - Should lead to recipe details.
   - f the user is the owner of the recipe, should allow the user to delete it.
- [x] Public recipe list:
   - Should display a list of all public recipes ordered by newest as in the wireframe.
   - Should lead to recipe details.
- [x] Recipe details:
   - If the recipe is public or the user is the owner of the recipe, should display the recipe details as in the wireframe.
   - If the user is the owner of the recipe, should lead to the form that allows the user to add new food.
-[x] General shopping list view (only for a group that has less than 3 members):
   - Should show the list of food that is missing for all recipes of the logged-in user (compare the list of food for all recipes with the general food list of that user).
   - Should count the total food items and total price of the missing food.
- [x] Make sure there are no N+1 queries happening.
- [x] Create a navigation menu that allows users to open all of the pages you created.
- [x] Write unit and integration tests

### CODE REVIEW

File: app/models/ability.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class Ability
+  include CanCan::Ability
+
+  def initialize(user)
+    user ||= User.new
+    if user.admin?
+      can :manage, :all
+    else
```
Comment (omarsalem7) [inline app/models/ability.rb:8]: - [ ] Kindly consider adding another rule for the only owner can delete an item(food, recipe) if it is his own which has already been created

**Review by omarsalem7 — CHANGES_REQUESTED** [review]
Hi @elarhadu, @houdamzari 😊, 

Good job so far! 👏
You have done well reaching this point of the milestone, but there are some issues that you still need to work on to go to the next project but you are almost there! 

you-are-near

## Highlights
- The design matches the wireframe ✔️
-  The app schema looks similar to the Entity Relationship Diagram✔️

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
- N/A

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/recipes/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+</div>
+
+  <div>
+    <% @recipes.each do |recipe| %>
+      <%= link_to recipe_path(recipe), class: "text-decoration-none text-dark" do %>
+          <%= render recipe %>
+      <% end %>
+    <% end %>
+  </div>
+
```
Comment (prolajumokeoni) [inline app/views/recipes/index.html.erb:14]: - [ ] Kindly make sure a logged-in user, can add and delete their own recipes, currently, I see this error when I try to add a recipe, please check the requirements and ensure that you implement the functionalities  

Screenshot_5

**Review by prolajumokeoni — CHANGES_REQUESTED** [review]
Hi @Team

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

## Highlights
✔️ Design matches wireframe
✔️ Application has a login and Registration page, built with the Devise gem

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification @prolajumokeoni.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/recipes/index.html.erb (hunk 2 above)
Comment (houdamzari) [reply app/views/recipes/index.html.erb:14]: @prolajumokeoni  Kindly fill the user input with the user id not the name .

Comment (prolajumokeoni) [reply app/views/recipes/index.html.erb:14]: Hi @houdamzari thanks for pointing this out, please ensure that after successfully adding a  recipe only the owner of the recipe can delete it 👍🏿

Comment (anagudelogu) [reply app/views/recipes/index.html.erb:14]: You did a great job working on this project! 🚀 However, there's still room to improve! I think what the previous reviewer meant was that the recipe should be created on behalf of the currently logged-in user, take in mind that regular users don't have access to the user's id as it is generated by the system.

- [ ] Please consider removing the User field and, instead, use the `current_user` method provided by devise to create a recipe on behalf of the current user.

File: app/views/shopping_list/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+              <span><%= total_quantity %> <%= ingredients.first.food.measurement_unit %></span>
+          </td>
+          <td>
+              <% total_price = ingredients.map { |ingredient| ingredient.quantity * ingredient.food.price }.sum %>
+              $<%= total_price %>
+          </td>
+        </tr>
+      <% end %>
+      </tbody>
+    </table>
```
Comment (anagudelogu) [inline app/views/shopping_list/index.html.erb:40]: - [ ] **[OPTIONAL]** Also, I have a take on the shopping list: I'm sorry that the previous reviewers didn't point this out so I'm marking this as optional to keep consistency between reviews, but according to the requirements:

image

The goal of this general shopping list is to list the missing food for all the recipes of the user, so for example, if the user has 5kg of `food1`, and one recipe asks for 10kg, the shopping list should show 5kg and not 10kg, which is what the user needs to buy in order to have enough for the recipe, take in mind that it should list also ingredients from other recipes if there's more than one! 🤓 🔧  

💡 Hint: Compare the `recipe_foods` quantity for the user's recipes with the user's `foods` quantity.

Currently, your app only takes the `recipe_foods` quantity and prints this in the shopping list, without taking into account the amount of food the user already has! I strongly suggest that you fix this! 🤓 🔧

**Review by anagudelogu — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED ♻️ 🚧 

Hi @elarhadu & @houdamzari 👋🏼 
I'm Andres.

⚠️ **PLEASE AVOID PUSHING NEW COMMITS TO YOUR PROJECT ONCE YOU SUBMITTED IT FOR REVIEW, ALWAYS MAKE SURE THAT YOUR PROJECT IS READY WHEN YOU REQUEST A REVIEW.** ⚠️ 

You've done a great job implementing some of the requested changes for this milestone 🚀, but there are still some which haven't been addressed yet. You're getting closer, keep up the good work! 🤖 🎆 



 ## Highlights
✅ Great use of Git flow. 💯 
✅ Your project is documented professionally. 📜 
✅ Great UI design. 💻 
✅ Descriptive PR title, summary and commit messages. 💬 


 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/recipes_controller.rb [hunk 4]
```diff
@@ -14,9 +14,11 @@ def new
 
   def create
     @recipe = Recipe.new(recipe_params)
+    @recipe.user_id = current_user.id
+
```
Comment (TadesseDev) [inline app/controllers/recipes_controller.rb]: - Good job Team :heavy_check_mark:

**Review by TadesseDev — APPROVED** [review]
### Hi Team

### I'm Tadesse Alemayehu your reviewer for this PR. :hand: 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 
giphy (1)



## Highlights
 - [x] Good job making the changes requested by the previous code reviewer. ✅

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**